render-ws-spark-client/src/main/java/org/janelia/render/client/spark/intensityadjust/ShadingCorrectionClient.java

```diff
@@ -0,0 +1,328 @@
+                if (model == null) {
+                    LOG.warn("No model found for z={}", z);
+                    continue;
+                }
+
+                final RandomAccessibleInterval<UnsignedShortType> imgSlice = Views.hyperSlice(croppedImg, 2, z - (int) block.min(2));
+                final RandomAccessibleInterval<UnsignedByteType> maskSlice = Views.hyperSlice(croppedMask, 2, z - (int) block.min(2));
+                final Cursor<UnsignedShortType> imgCursor = Views.iterable(imgSlice).localizingCursor();
+                final Cursor<UnsignedByteType> maskCursor = Views.iterable(maskSlice).localizingCursor();
+                final double[] location = new double[2];
+
+                // apply model to slice
+                while (imgCursor.hasNext()) {
+                    final UnsignedShortType pixel = imgCursor.next();
+                    final UnsignedByteType maskPixel = maskCursor.next();
+                    if (maskPixel.get() == 0) {
+                        // don't apply model to background
+                        continue;
+                    }
+
+                    location[0] = ShadingModel.scaleCoordinate(imgCursor.getDoublePosition(0), xScale);
+                    location[1] = ShadingModel.scaleCoordinate(imgCursor.getDoublePosition(1), yScale);
+
+                    model.applyInPlace(location);
+                    pixel.setReal(UnsignedShortType.getCodedSignedShortChecked((int) (pixel.getRealDouble() - location[0])));
+                }
+            }
+
+            // save block
+            N5Utils.saveNonEmptyBlock(croppedImg, out, parameters.datasetOut, block.gridPosition, new UnsignedShortType());
+        }
+    }
+
+
+    static class ShadingModelProvider implements Serializable {
+        private final List<ModelSpec> sortedModelSpecs;
+
+        private ShadingModelProvider(final List<ModelSpec> modelSpecs) {
+            this.sortedModelSpecs = modelSpecs;
+            this.sortedModelSpecs.sort(Collections.reverseOrder(Comparator.comparingInt(ModelSpec::getZ)));
+        }
+
+        public ShadingModel getModel(final int z) {
+            for (final ModelSpec modelSpec : sortedModelSpecs) {
+                if (z >= modelSpec.getZ()) {
+                    return modelSpec.getModel();
+                }
+            }
+            return null;
+        }
+
+        public static ShadingModelProvider fromJsonFile(final String fileName) throws IOException {
+            LOG.info("Reading model specs from file: {}", fileName);
+            try (final FileReader reader = new FileReader(fileName)) {
+                return fromJson(JsonParser.parseReader(reader));
+            }
+        }
+
+        public static ShadingModelProvider fromJson(final JsonElement jsonData) {
+
+            final List<ModelSpec> modelSpecs = new ArrayList<>();
+            Collections.addAll(modelSpecs, new Gson().fromJson(jsonData, ModelSpec[].class));
```

Comment:
To avoid problems with ancient Gson libraries deployed with Spark, I usually use Jackson libraries for JSON stuff (see org.janelia.alignment.json.JsonUtils).  The N5 libraries already depend upon Gson  - so use of it here does not make things worse.  Just wanted to let you know the history/rationale ...

Reply:
That's actually good to know. I just went with gson because it was easier to make the parsing work. But co-pilot had no problems translating, so I'll use Jackson.